# Tarea 03 — Procesamiento de Datos con Pandas

**Instrucciones:** Completa cada celda de código con tu respuesta.  
Trabajarás con el dataset del Titanic para practicar transformaciones reales.

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


---
## Parte 1 — Exploración y selección

**P1.** Imprime los nombres de todas las columnas y el número total de filas y columnas del DataFrame.

In [3]:
print(df.columns.tolist())
print(df.shape)

['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']
(891, 12)


**P2.** Usa `.loc` para seleccionar las filas 10 a 15 (inclusive) y solo las columnas `Name`, `Age` y `Survived`.

In [5]:
seleccion = df.loc[10:15, ["Name", "Age", "Survived"]]

print(seleccion)


                                    Name   Age  Survived
10       Sandstrom, Miss. Marguerite Rut   4.0         1
11              Bonnell, Miss. Elizabeth  58.0         1
12        Saundercock, Mr. William Henry  20.0         0
13           Andersson, Mr. Anders Johan  39.0         0
14  Vestrom, Miss. Hulda Amanda Adolfina  14.0         0
15      Hewlett, Mrs. (Mary D Kingcome)   55.0         1


---
## Parte 2 — Valores nulos y limpieza

**P3.** Cuenta cuántos valores nulos hay en cada columna. ¿Cuáles columnas tienen nulos?

In [6]:
nulos_por_columna = df.isnull().sum()

print("Cantidad total de nulos por columna:")
print(nulos_por_columna)

print("\n---")

columnas_con_nulos = nulos_por_columna[nulos_por_columna > 0]

print("Las columnas que tienen valores nulos son:")
print(columnas_con_nulos)


Cantidad total de nulos por columna:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

---
Las columnas que tienen valores nulos son:
Age         177
Cabin       687
Embarked      2
dtype: int64


**Respuesta:** Las columnas con valores nulos son: Age, Cabin y Embarked

**P4.** Rellena los valores nulos de la columna `Age` con la **mediana** de esa columna.  
Verifica que ya no haya nulos en `Age`.

In [7]:
mediana_edad = df['Age'].median()
print("La mediana de edad es:", mediana_edad)

df['Age'] = df['Age'].fillna(mediana_edad)

nulos_edad_ahora = df['Age'].isnull().sum()

print("Cantidad de nulos en Age después de rellenar:", nulos_edad_ahora)


La mediana de edad es: 28.0
Cantidad de nulos en Age después de rellenar: 0


---
## Parte 3 — Nuevas columnas y agrupaciones

**P5.** Crea una nueva columna llamada `FamilySize` que sea la suma de `SibSp` + `Parch` + 1 (el pasajero mismo).

In [9]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

print(df[['SibSp', 'Parch', 'FamilySize']].head())


   SibSp  Parch  FamilySize
0      1      0           2
1      1      0           2
2      0      0           1
3      1      0           2
4      0      0           1


**P6.** Usando `groupby`, calcula la **tasa de supervivencia promedio** (columna `Survived`) por clase (`Pclass`).  
¿Qué clase tuvo mayor probabilidad de sobrevivir?

In [12]:
tasa_supervivencia = df.groupby('Pclass')['Survived'].mean()

print(tasa_supervivencia)


Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


**Respuesta:** La clase con mayor supervivencia fue:  La primera clase (1)

**P7.** Calcula la edad promedio y la tarifa (`Fare`) promedio, agrupando por `Sex` y `Pclass`.  
Usa `.groupby()` con `.agg()`.

In [13]:
agrupacion = df.groupby(['Sex', 'Pclass'])

promedios = agrupacion.agg({
    'Age': 'mean', 
    'Fare': 'mean'
})

print(promedios)

                     Age        Fare
Sex    Pclass                       
female 1       33.978723  106.125798
       2       28.703947   21.970121
       3       23.572917   16.118810
male   1       38.995246   67.226127
       2       30.512315   19.741782
       3       26.911873   12.661633


---
## Parte 4 — Ordenamiento y duplicados

**P8.** Ordena el DataFrame por `Fare` de mayor a menor y muestra las primeras 5 filas.  
¿Cuánto pagó el pasajero más caro?

In [15]:
df_ordenado = df.sort_values(by='Fare', ascending=False)

print(df_ordenado[['Name', 'Pclass', 'Fare']].head(5))

                                   Name  Pclass      Fare
679  Cardeza, Mr. Thomas Drake Martinez       1  512.3292
258                    Ward, Miss. Anna       1  512.3292
737              Lesurer, Mr. Gustave J       1  512.3292
88           Fortune, Miss. Mabel Helen       1  263.0000
438                   Fortune, Mr. Mark       1  263.0000


**Respuesta:** Hubo 3 pasajeros que pagaron el pasaje más caro que fue por un valor de 512.3292

**P9.** Verifica si hay filas duplicadas en el DataFrame. ¿Cuántas hay?

In [17]:
cantidad_duplicados = df.duplicated().sum()

print(cantidad_duplicados)

0


**Respuesta:** No hay filas duplicadas